### PR TITLE
initial gRPC bulk publish impl

### PR DESCRIPTION
**Pull request by mukundansundar**
# Description

initial gRPC bulkPublish change. without E2E

to be merged after https://github.com/dapr/dapr/pull/5127 is merged.

This fixes HTTP E2E which were only testing the default impl and not kafka impl .... 
This adds gRPC E2E ....

## Issue reference

related to #2218 

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: pkg/grpc/api_test.go [hunk 1]
```diff
@@ -1797,6 +1801,21 @@ func TestPublishTopic(t *testing.T) {
 			GetPubSubFn: func(pubsubName string) pubsub.PubSub {
 				return &daprt.MockPubSub{}
 			},
+			BulkPublishFn: func(req *pubsub.BulkPublishRequest) (pubsub.BulkPublishResponse, error) {
+				if req.Topic == "error-topic" {
```
Comment (pravinpushkar) [inline pkg/grpc/api_test.go:1805]: nit: switch case here

File: pkg/http/api.go [hunk 2]
```diff
@@ -2020,6 +1998,221 @@ func (a *api) onPublish(reqCtx *fasthttp.RequestCtx) {
...
 
+type bulkPublishMessageEntry struct {
+	EntryID     string            `json:"entryID,omitempty"`
+	Event       interface{}       `json:"event"`
+	ContentType string            `json:"contentType"`
+	Metadata    map[string]string `json:"metadata,omitempty"`
+}
+
+func (a *api) onBulkPublish(reqCtx *fasthttp.RequestCtx) {
+	thepubsub, pubsubName, topic, sc, msg := a.validateAndGetPubsubAndTopic(reqCtx)
```
Comment (pravinpushkar) [inline pkg/http/api.go:2009]: nit: msg to be errMsg/errResponse ?

Comment (mukundansundar) [reply pkg/http/api.go:2009]: fixed in http bulk publish PR....

File: go.mod [hunk 3]
```diff
@@ -394,6 +394,10 @@ replace github.com/eclipse/paho.mqtt.golang => github.com/shivamkm07/paho.mqtt.g
 // Don't commit with this uncommented!
 //
 // replace github.com/dapr/components-contrib => ../components-contrib
+
 //
 // Then, run `make modtidy` in this repository.
 // This ensures that go.mod and go.sum are up-to-date.
+
+// This is the feature/pubsub-batching branch in components-contrib, will be removed before merging into master branch of dapr
+replace github.com/dapr/components-contrib => github.com/dapr/components-contrib v1.8.0-rc.1.0.20220920200252-e9e750bce615
```
Comment (daixiang0) [inline go.mod:403]: Is it planned to be merged this week? I do not want to introduce replace here.

File: pkg/diagnostics/component_monitoring.go [hunk 4]
```diff
@@ -159,6 +174,30 @@ func (c *componentMetrics) PubsubIngressEvent(ctx context.Context, component, pr
...
 
+// BulkPubsubEgressEvent records the metris for a pub/sub egress event.
+// eventCount if greater than zero implies successful publish of few/all events in the bulk publish call
+func (c *componentMetrics) BulkPubsubEgressEvent(ctx context.Context, component, topic string, success bool, eventCount int64, elapsed float64) {
+	if c.enabled {
+		stats.RecordWithTags(
+			ctx,
+			diagUtils.WithTags(appIDKey, c.appID, componentKey, component, namespaceKey, c.namespace, successKey, fmt.Sprintf("%v", success), topicKey, topic),
+			c.bulkPubsubEgressCount.M(1))
+		if eventCount > 0 {
```
Comment (daixiang0) [inline pkg/diagnostics/component_monitoring.go:185]: Do we have tests for this case?

File: pkg/http/api.go [hunk 5]
```diff
@@ -2020,6 +1998,221 @@ func (a *api) onPublish(reqCtx *fasthttp.RequestCtx) {
...
+			}
+			bulkRes.Statuses[i].Status = string(r.Status)
+		}
+	}
+	diag.DefaultComponentMonitoring.BulkPubsubEgressEvent(context.Background(), pubsubName, topic, err == nil, eventsPublished, elapsed)
+	if err != nil {
+		status := fasthttp.StatusInternalServerError
+		bulkRes.ErrorCode = "ERR_PUBSUB_PUBLISH_MESSAGE"
+
+		if errors.As(err, &runtimePubsub.NotAllowedError{}) {
```
Comment (daixiang0) [inline pkg/http/api.go:2160]: Could we use switch case here to reuse code?

File: pkg/http/api.go [hunk 6]
```diff
@@ -2020,6 +1998,221 @@ func (a *api) onPublish(reqCtx *fasthttp.RequestCtx) {
...
+		if errors.As(err, &runtimePubsub.NotFoundError{}) {
+			msg := NewErrorResponse("ERR_PUBSUB_NOT_FOUND", err.Error())
+			status = fasthttp.StatusBadRequest
+			respond(reqCtx, withError(status, msg), closeChildSpans)
+			log.Debug(msg)
+
+			return
+		}
+
+		resData, _ := json.Marshal(bulkRes)
```
Comment (daixiang0) [inline pkg/http/api.go:2178]: respond code can be merged.

File: go.mod (hunk 3 above)
Comment (mukundansundar) [reply go.mod:403]: For sure ... it will be resolved before a PR is created to master ...

File: pkg/diagnostics/component_monitoring.go (hunk 4 above)
Comment (mukundansundar) [reply pkg/diagnostics/component_monitoring.go:185]: Will see if I can add a test for this.

File: pkg/runtime/runtime.go [hunk 7]
```diff
@@ -1792,6 +1792,23 @@ func (a *DaprRuntime) Publish(req *pubsub.PublishRequest) error {
...
+	}
+
+	if allowed := a.isPubSubOperationAllowed(req.PubsubName, req.Topic, ps.scopedPublishings); !allowed {
+		return pubsub.BulkPublishResponse{}, runtimePubsub.NotAllowedError{Topic: req.Topic, ID: a.runtimeConfig.ID}
+	}
+	if bulkPublisher, ok := ps.component.(pubsub.BulkPublisher); ok {
+		return bulkPublisher.BulkPublish(context.TODO(), req)
+	}
+
+	log.Warnf("pubsub %s does not implment the bulkPublish API, defaulting to normal publish", req.PubsubName)
```
Comment (artursouza) [inline pkg/runtime/runtime.go:1808]: Log debug and fix typo, please.

File: tests/apps/pubsub-publisher/app.go [hunk 8]
```diff
@@ -107,13 +108,22 @@ func getBulkRequetMetadata(r *http.Request) map[string]string {
 	return metadata
 }
 
+func getBulkPublishPubsubOrDefault(reqMeta map[string]string) string {
+	if v, ok := reqMeta[bulkPubsubMetaKey]; ok {
+		delete(reqMeta, bulkPubsubMetaKey)
+		return v
+	}
+	return pubsubName
+}
```
Comment (shubham1172) [inline tests/apps/pubsub-publisher/app.go:117]: This function is doing something different than its name, no?

Comment (mukundansundar) [reply tests/apps/pubsub-publisher/app.go:117]: Can be changed to getAndDeleteBulkPubsubnameOrDefault....felt that the name was too long...

Will add a comment for the function as well as where it is being called that the metadata for the test is being deleted

**Comment by artursouza** [comment]
Merging into feature branch. Let's do a deep review from feature into master PR.

**Pull request merged**